docs: add `dataProvider` section on new tutorial (#3305)
refinedev/refine: documentation/docs/tutorial/2-understanding-dataprovider/0-intro.md

---
id: index
title: 1. Data Provider
tutorial:
    order: 0
    prev: false
    next: tutorial/understanding-dataprovider/swizzle
---

:::info
DataProvider unit is not required for the tutorial. You can skip this unit if you want.
Go to the next unit [Adding CRUD Pages](#) to continue.
:::

## What is data provider?

The dataProvider acts as a data layer for your app that makes the HTTP requests and encapsulates how the data is retrieved. **refine** consumes these methods via data hooks.

refine comes with different data providers out of the box, but the one we’re interested in and will be using in this tutorial is the refine-simple-rest data provider, a data provider for communicating with RESTful APIs.

:::tip
Browse the [list of supported providers](#supported-data-providers) to select an open source package for your API.
You can also check the details to [create your own data provider](/docs/tutorial/understanding-dataprovider/create-dataprovider/) here.
:::

> The data provider concept comes into play whenever your refine app needs to communicate with an external API. Thanks to refine's `dataProvider` object property, all API requests are processed through this object.

Reviewer: We can move this sentence to line 18

Data providers can communicate with `REST`, `GraphQL`, `RPC` and `SOAP` based APIs as they work with adapter system infrastructure.

*TODO: Maybe you draw some diagrams to visualize it?*

Data Provider must have the following methods:

```ts

> import { DataProvider } from "@pankod/refine-core";

Reviewer: Where do you use this import?

Author: Use `dataProvider` return type. `const dataProvider: DataProvider = {`

const dataProvider: IDataProvider = {
    create: ({ resource, variables, metaData }) => Promise,
    deleteOne: ({ resource, id, variables, metaData }) => Promise,
    getList: ({
        resource,
        pagination,
        hasPagination,
        sort,
        filters,
        metaData,
    }) => Promise,
    getMany: ({ resource, ids, metaData }) => Promise,
    getOne: ({ resource, id, metaData }) => Promise,
    update: ({ resource, id, variables, metaData }) => Promise,
    getApiUrl: () => "",
}
```

> 

Reviewer: You can add something like: refine uses this method internally to perform CRUD operations..

## Using Data Providers in refine

You can pass your data provider to the `<Refine>` component as a prop.

```tsx
import dataProvider from "@pankod/refine-simple-rest"

<Refine 
    dataProvider={dataProvider("https://api.fake-rest.refine.dev")}
/>
```
You can refer to the **refine** component [dataProvider](/docs/api-reference/core/components/refine-config/#dataprovider) prop documentation for more detailed information.

## How is it work with data hooks?

> **refine** data hooks are used to fetch data from the data provider. It passed paramters to the data provider and returns the result.

Reviewer: ```suggestion
**refine** data hooks are used to fetch data from the API using the data provider methods. It passes parameters to the data provider and returns the result.
```

For example, to get all the records for the `posts` resource, use the `useList` hook as below. 

```ts title="src/pages/posts/index.tsx"
import { useList } from "@pankod/refine-core";

const postUseListResult = useList({
    resource: "posts",
    config: {
      sort: [
        {
          field: "id",
          order: "desc",
        },
      ],
      filters: [
        {
          field: "title",
          operator: "contains",
          value: "hello",
        },
      ],
    },
  });
```

> 

Reviewer: You need to say something like: In ht background refine's useList hook uses getList from dataProvider..

For example, to get all the records for the `posts` resource, use the `useList` hook as below. 

```ts title="src/pages/posts/index.tsx"
import { useList } from "@pankod/refine-core";

const postUseListResult = useList({
    resource: "posts",
    config: {
      sort: [
        {
          field: "id",
          order: "desc",
        },
      ],
      filters: [
        {
          field: "title",
          operator: "contains",
          value: "hello",
        },
      ],
    },
  });
```

```ts title="dataProvider.ts"
const dataProvider = {
    getList: (params) => {
        console.log(params);
        /*
        {
          "resource": "posts",
          "sort": [
            {
              "field": "id",
              "order": "desc"
            }
          ],
          "filters": [
            {
              "field": "title",
              "operator": "contains",
              "value": "hello"
            }
          ],
        }
        */
    }
    ...
}
```

## Supported Data Providers

**refine** includes many out-of-the-box data providers to use in your projects like 

> 

Reviewer: You can explain that how to use these data providers. like you can install afterwards or selecting when bootstrapping a project with cli...

---
id: index
title: 1. Data Provider
tutorial:
    order: 0
    prev: false
    next: tutorial/understanding-dataprovider/swizzle
---

:::info

> The DataProvider unit is optional for the tutorial and can be skipped if desired.

Reviewer: ```suggestion
The DataProvider unit is optional for the tutorial and can be skipped to next unit [Adding CRUD Pages](#) if desired.
```

> Go to the next unit [Adding CRUD Pages](#) to continue.

Reviewer: ```suggestion

```

:::

## What is data provider?

> The dataProvider acts as a data layer for your app that makes the HTTP requests and encapsulates how the data is retrieved. **refine** consumes these methods via data hooks.

Author: Data Provider methods

The data provider concept comes into play whenever your refine app needs to communicate with an external API. Thanks to refine's `dataProvider` object property, all API requests are processed through this object.

> 

Reviewer: You don't need worry about creating data providers from scratch . refine offers built-in data provider support for the most popular API providers(link to supported data providers). So you can use one of them or you can create your own data provider according to your needs. We 'll see how to create data provider in the next sections.

refine comes with different data providers out of the box, but the one we’re interested in and will be using in this tutorial is the refine-simple-rest data provider, a data provider for communicating with RESTful APIs.

:::tip
Browse the [list of supported providers](#supported-data-providers) to select an open source package for your API.
You can also check the details to [create your own data provider](/docs/tutorial/understanding-dataprovider/create-dataprovider/) here.
:::

Data providers can communicate with `REST`, `GraphQL`, `RPC` and `SOAP` based APIs as they work with adapter system infrastructure.

<br/>
<br/>
<div>
    <img src="https://refine.ams3.cdn.digitaloceanspaces.com/website/static/img/guides-and-concepts/providers/data-provider/flow.png" />
</div>
<br/>
<br/>

> Data Provider must have the following methods:

Reviewer: ```suggestion
The typical data provider has following methods:
```

---
id: index
title: 1. Data Provider
tutorial:
    order: 0
    prev: false
    next: tutorial/understanding-dataprovider/swizzle
---

:::info
The data provider unit is optional for the tutorial and can be skipped to next unit [Adding CRUD Pages](/docs/tutorial/adding-crud-pages/antd/index) if desired.
:::

## What is data provider?

The data provider acts as a data layer for your app that makes the HTTP requests and encapsulates how the data is retrieved. **refine** consumes these methods via data hooks.

You don't need worry about creating data providers from scratch. **refine** offers built-in data provider support for the most popular [API providers](#supported-data-providers). So you can use one of them or you can create your own data provider according to your needs. We'll see how to create data provider in the next sections.

:::tip

> You can review the[list of supported providers](#supported-data-providers) to find the one that is compatible with your API. 

Reviewer 2: ```suggestion
You can review the [list of supported providers](#supported-data-providers) to find the one that is compatible with your API. 
```

<div>
    <img src="https://refine.ams3.cdn.digitaloceanspaces.com/website/static/img/guides-and-concepts/providers/data-provider/flow.png" />
</div>
<br/>
<br/>

The typical data provider has following methods:

```ts
import { DataProvider } from "@pankod/refine-core";

const dataProvider: DataProvider = {
    create: ({ resource, variables, metaData }) => Promise,
    deleteOne: ({ resource, id, variables, metaData }) => Promise,
    getList: ({
        resource,
        pagination,
        hasPagination,
        sort,
        filters,
        metaData,
    }) => Promise,
    getMany: ({ resource, ids, metaData }) => Promise,
    getOne: ({ resource, id, metaData }) => Promise,
    update: ({ resource, id, variables, metaData }) => Promise,
    getApiUrl: () => "",
}
```

These methods are used to perform CRUD operations by **refine**

**refine** comes with different data providers out of the box, but the one we’re interested in and will be using in this tutorial is the refine-simple-rest data provider, a data provider for communicating with RESTful APIs.

> ## Using Data Providers in refine

Author: `You can refer to the refine component [dataProvider](http://localhost:3000/docs/api-reference/core/components/refine-config/#dataprovider) prop documentation for more detailed information.` We've added this, any idea what else we can do? @necatiozmen

You can register your data provider to the `<Refine>` component by passing to `dataProvider` property.

```tsx
import dataProvider from "@pankod/refine-simple-rest"

<Refine 
    dataProvider={dataProvider("https://api.fake-rest.refine.dev")}
/>
```
You can refer to the **refine** component [dataProvider](/docs/api-reference/core/components/refine-config/#dataprovider) prop documentation for more detailed information.

## How data provider methods are using in the app?

> We use refine's data hooks whenever we need to fetch data from the API. These data hooks are connected to data provider methods internally. The required parameters passes to the data provider methods and response from API returning back.

Reviewer 2: ```suggestion
We use **refine's** data hooks whenever we need to fetch data from the API. These data hooks are connected to data provider methods internally. The required parameters passes to the data provider methods and response from API returning back.
```

> We use **refine's** data hooks whenever we need to fetch data from the API. These data hooks are connected to data provider methods internally. The required parameters passes to the data provider methods and response from API returning back.

Reviewer: ```suggestion
We use **refine's** data hooks whenever we need to fetch data from the API. These data hooks are connected to data provider methods internally. The required parameters passes to the data provider methods and response from API returning back.  

```

> To illustrate this internal connection, imagine we want to get all records from the "post" resource using refine's `useList` data hook.

Reviewer: ```suggestion
To illustrate this internal connection, imagine we want to get all records from the `post` resource using refine's `useList` data hook.
```

```ts title="src/pages/posts/index.tsx"
import { useList } from "@pankod/refine-core";

const postUseListResult = useList({
    resource: "posts",
    config: {
      sort: [
        {
          field: "id",
          order: "desc",
        },
      ],
      filters: [
        {
          field: "title",
          operator: "contains",
          value: "hello",
        },
      ],
    },
  });
```

> As we can see, the parameters passed to useList hook are forwarded to the data providers getList method internally. In the background, refine connects all data provider methods to appropriate data hooks.

Reviewer 2: ```suggestion
As we can see, the parameters passed to the `useList` hook are forwarded to the data provider's `getList` method internally. In the background, **refine** connects all data provider methods to appropriate data hooks.
```

```ts title="dataProvider.ts"
const dataProvider = {
    getList: (params) => {
        console.log(params);
        /*
        {
          "resource": "posts",
          "sort": [
            {
              "field": "id",
              "order": "desc"
            }
          ],
          "filters": [
            {
              "field": "title",
              "operator": "contains",
              "value": "hello"
            }
          ],
        }
        */
    }
    ...
}
```

## Supported Data Providers

> **refine** comes with many pre-loaded data providers. To include them in your project, you can use `npm install [packageName]` or you can use the `npm create refine-app@latest projectName` to configure them. This will allow you to easily use these data providers in your project.

Reviewer: ```suggestion
**refine** comes with many pre-loaded data providers. To include them in your project, you can use `npm install [packageName]` or you can select preferred data provider with the `npm create refine-app@latest projectName` during the project creation phase with CLI.  
 This will allow you to easily use these data providers in your project.
```

---
id: index
title: 1. Data Provider
tutorial:
    order: 0
    prev: false
    next: tutorial/understanding-dataprovider/swizzle
---

 ## What is data provider?

> When it needs to communicate with the *refine* API, it does so using the `dataProvider` object. All API requests are processed through this object.

Reviewer: ```suggestion
The data provider concept comes into play whenever your refine app needs to communicate with an external API. Thanks to refine's `dataProvider` object property,  all API requests are processed through this object.
```

> 

Reviewer: You need to connect this section to the previous one.Readers have seen the App.tsx and the `dataProvider` property in the previous section. This is where we need to explain   <Refine> components(in App.tsx)  `dataProvider` property to the users.
